string path
---
int32[] xs
int32[] zs
